### PR TITLE
Small improvements to the API

### PR DESCRIPTION
**Pull request by FrankHeijden**
- Makes Claim#getChunkHashes() accessible to the outside world
- Implements Claim#clone()
- Refactors Visualisation to allow claim visualisations from outside of GP, with custom blocks.
- Adds a shift-inspecting flag to VisualizationEvent, to check if the event was called when a player shift-clicked with the inspection tool.

### CODE REVIEW

**Comment by FrankHeijden** [comment]
This also closes #405

**Comment by Jikoo** [comment]
* Why do you need `Claim#getChunkHashes`? `DataStore#getChunkClaims` handles that for you, all you would need is the chunks containing your extra border.
* Visualization changes: your diff is unreadable because you reformatted at the same time as making changes you needed. Those should ideally be two commits because so much changed - GH can't figure out anything similar, and I as a mobile user without PC access sure as heck am struggling to see the actual diff.

**Comment by FrankHeijden** [comment]
Apparently commits to `Visualisation` bork the diff. Basically only change happened there was the split of addClaimElements: https://github.com/TechFortress/GriefPrevention/pull/937/commits/d6dd5ef7f2dd17899dca4e86109186212b27f6be#diff-4bfa3722c20a50287e14a4c271a27947R135

And the change of how we check for “outside” elements: https://github.com/TechFortress/GriefPrevention/pull/937/commits/d6dd5ef7f2dd17899dca4e86109186212b27f6be#diff-4bfa3722c20a50287e14a4c271a27947R234

That last one is needed because the method shouldn’t depend on a claim input, but instead of a more general cuboid (represented by `Location min` and `Location max`)

**Comment by FrankHeijden** [comment]
> Why do you need Claim#getChunkHashes? DataStore#getChunkClaims handles that for you, all you would need is the chunks containing your extra border.

@Jikoo Sure could be, but I prefer it this way to quickly lookup claims based on their AntiClaimZone (thus storing the anticlaimzone in the hashmap just like GP does with normal claims), without additionally needing to check with the GP api if the chunk was an actual claimchunk. The overhead of doing this is not worth it versus just opening up a method to the outside world.

**Comment by FrankHeijden** [comment]
Guess GitHub is just broken on that file, lol

**Comment by Jikoo** [comment]
Re: Visualization - Check your line endings, usually a full file diff is a result of \n\r <-> \n

> Sure could be, but I prefer it this way to quickly lookup claims based on their AntiClaimZone, without additionally needing to check with the GP api if the chunk was an actual claimchunk. The overhead of doing this is not worth it versus just opening up a method to the outside world.

I see, you want to piggyback your own storage on GP's hashing system. In that case, you should probably create a utility class devoted to related static methods surrounding this instead (i.e. also migrate `DataStore#getChunkHash` and similar). If the multi-chunk method is exposed the single chunk method should be too. Would like to see that as a separate PR though, it will be a decent-sized diff.

**Comment by FrankHeijden** [comment]
What exactly are you asking me to do right now? should I revert the `getChunkHashes()` part? I really cannot see why we can't have the method like it is right now in the PR haha. it would not break existing implementations, nor is it a bad thing to "expose" 😉. But if it needs to be yeeted, I'll make a commit for it.

**Comment by Jikoo** [comment]
> What exactly are you asking me to do right now? should I revert the `getChunkHashes()` part? I really cannot see why we can't have the method like it is right now in the PR haha. it would not break existing implementations, nor is it a bad thing to "expose" 😉. But if it needs to be yeeted, I'll make a commit for it.

I don't think it's a bad addition to allow addons to calculate chunk hashes the way GP does, but I think the way you're doing it doesn't make sense from a design perspective.
It's not just method exposure, you're adding a new utility overload that could be static. If you're adding a static utility method, it makes sense to group it with all related static utility methods. If an addon needs to calculate multiple chunk hashes, what are the odds that the same addon will never want to calculate a single chunk's hash? Your particular use case doesn't need it, but any new changes are semi-permanent API additions and should be a mature design. `Claim#getHashes` is an appropriate helper, but calculating arbitrary bounds' hashes is not. Maybe it belongs in DataStore with an exposed `getChunkHash`?

**Comment by FrankHeijden** [comment]
> you're adding a new utility overload that could be static

@Jikoo The method is static, but I can surely move the method over to dataStore.

File: src/main/java/me/ryanhamshire/GriefPrevention/CreateClaimResult.java [hunk 1]
```diff
@@ -21,9 +21,15 @@
...
+    public Value value;
 
     //when succeeded, this is a reference to the new claim
     //when failed, this is a reference to the pre-existing, conflicting claim
     public Claim claim;
+
+    public enum Value {
+        SUCCEEDED,
+        FAILED,
+        CANCELLED
```
Comment (RoboMWM) [inline src/main/java/me/ryanhamshire/GriefPrevention/CreateClaimResult.java:33]: Where is this used...?

**Comment by RoboMWM** [comment]
As you may already know I'm not a fan of one PR implementing multiple changes, and since this is a more mature and depended-on plugin I want to be sure what changes this is making to the plugin.

I'm also not really a fan of the changes here. dev/v20 can give you some ideas on what I want to change in this plugin; things like chunkhash calculations are moved to a separate class (iirc). And ya, diff for Visualization screwed up entirely.

Not sure what the benefit of claim#clone is, nor am I aware of it being a good practice (vs. just reusing the constructor)

File: src/main/java/me/ryanhamshire/GriefPrevention/CreateClaimResult.java (hunk 1 above)
Comment (FrankHeijden) [reply src/main/java/me/ryanhamshire/GriefPrevention/CreateClaimResult.java:33]: When cancelling the ClaimModifiedEvent, we don’t want GP to act on a cancelled event.

**Comment by FrankHeijden** [comment]
> I'm also not really a fan of the changes here.

@RoboMWM I really don’t see the issue here. The methods I have opened up are all backwards compatible with previous version of GP, and I will need these API methods for the described addon #937 to work.

> dev/v20 can give you some ideas on what I want to change in this plugin; things like chunkhash calculations are moved to a separate class (iirc).

When v20 comes out, I will update the addon. Since v20 is a major change, it does not matter if it contains breaking API. This PR will just ensure I can use the methods needed now.

> And ya, diff for Visualization screwed up entirely.

That’s something out of my bounds. I am using intellij idea, there it shows the diff correctly. Like Jikoo said above, it could be that the file ending was scuffed.

> Not sure what the benefit of claim#clone is, nor am I aware of it being a good practice (vs. just reusing the constructor)

All `.clone()` methods use the constructor. I am using this to make a copy of the “old” claim for ClaimModifiedEvent. It only makes sense that this event is an event which presents a “change”, with a “from” element and a “to”. Since this event is called in multiple pieces of the code, duplicating this is bad practise.

I am not adding useless methods. The changes in this PR are minor, and likely won’t affect any or hardly any addon out there. The changes will however make sure I can make my addon, and possibly many other addons, which was denied in GriefPrevention-core. 

Please give me concrete steps what you want changed and I will change it.

**Comment by FrankHeijden** [comment]
The problem with Visualization.java is indeed the CRLF file endings. There seems to be a mix of CRLF / LF ended files within the GP repo.

**Comment by FrankHeijden** [comment]
See #940, #941, #942 and #943.

**Comment by RoboMWM** [comment]
the v20 jump is no longer the plan, and instead we'll incrementally make it so.

**Pull request closed without merging**